spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java

```diff
@@ -63,7 +67,17 @@ protected MessageHandler createHandler(Object bean, Method method, List<Annotati
 					return new ReplyProducingMessageHandlerWrapper((MessageHandler) target);
 				}
 				else {
-					serviceActivator = new ServiceActivatingHandler(target);
+					ResolvableType methodReturnType = ResolvableType.forMethodReturnType(method);
+					if (Function.class.isAssignableFrom(methodReturnType.toClass()) ||
+							Consumer.class.isAssignableFrom(methodReturnType.toClass())) {
+
+						Class<?> expectedPayloadType = methodReturnType.getGeneric(0).toClass();
+						serviceActivator = new ServiceActivatingHandler(
+								new LambdaMessageProcessor(target, expectedPayloadType));
```

Comment:
I think you need to check for wildcard generic types here.

Reply:
Not sure what you mean, but with the `methodReturnType.getGeneric(0).toClass()` it returns an `Object` class for a `Consumer<?>`. So, I think we are good.